File: stdlib/js/hull/middleware/csp.js
/**
 * @file hull:middleware:csp
 * @module hull:middleware:csp
 * @description Content-Security-Policy middleware with per-request nonce.
 *
 * Lua parity: same surface as `hull.middleware.csp` with camelCase
 * options (`reportOnly` vs `report_only`).
 *
 * Generates a fresh 128-bit nonce on every request, exposes it via
 * `req.ctx.cspNonce`, and sets the `Content-Security-Policy` header
 * on the outbound response. Templates that need a nonce read it
 * from the context and emit it on `<script>` and `<style>` tags:
 *
 *     <script nonce="{{ csp_nonce }}">...</script>
 *     <style nonce="{{ csp_nonce }}">...</style>
 *
 * Note: the ctx key is `csp_nonce` (snake_case), not `cspNonce`. This
 * matches both the template literal and the JS `csrf` middleware's
 * snake-case `csrf_token` key, so apps can share templates across
 * the Lua and JS runtimes without naming churn.
 *
 * Three named profiles, identical to the Lua side:
 *
 *  - `csp.htmx()`. Pico-compatible (allows inline `style=` attributes
 *    via `style-src-attr 'unsafe-inline'`).
 *  - `csp.strict()`. No inline-style escape.
 *  - `csp.custom(opts)`. Caller-controlled directives.
 *
 * @license AGPL-3.0-or-later
 */

import { crypto } from "hull:crypto";

// Base64url-encode raw bytes (128-bit by default) for use as a CSP
// nonce. 16 bytes -> 22 base64url chars unpadded.
//
// JS-side note: crypto.random returns an ArrayBuffer (use .byteLength,
// not .length). crypto.base64urlEncode expects a JS string, so we
// build a binary-string view via Uint8Array + String.fromCharCode
// before encoding. Same pattern as stdlib/js/hull/jwt.js.
function nonceB64url(nBytes) {
    nBytes = nBytes || 16;
    const ab = crypto.random(nBytes);
    if (!ab || ab.byteLength !== nBytes) return null;
    const bytes = new Uint8Array(ab);
    let bin = "";
    for (let i = 0; i < bytes.length; i++) bin += String.fromCharCode(bytes[i]);
    const b64 = crypto.base64urlEncode(bin);
    if (!b64) return null;
    // Defensive: strip trailing '=' if the encoder ever returns padded.
    return b64.replace(/=+$/, "");
}

// Build the CSP header value for a given profile + nonce.
function buildHeader(profile, nonce) {
    const nonceToken = `'nonce-${nonce}'`;
    const parts = [
        "default-src 'self'",
        `script-src 'self' ${nonceToken}`,
        `style-src 'self' ${nonceToken}`,
        "img-src 'self' data:",
        "form-action 'self'",
        "frame-ancestors 'none'",
        "base-uri 'self'",
    ];
    if (profile === "htmx") {
        // Pico v2 classless uses inline style="..." attributes.
        // style-src-attr controls ONLY inline style attributes;
        // <style> blocks still require the nonce above.
        parts.push("style-src-attr 'unsafe-inline'");
    }
    return parts.join("; ");
}

function makeMiddleware(profile, opts) {
    opts = opts || {};
    const headerName = opts.reportOnly
        ? "Content-Security-Policy-Report-Only"
        : "Content-Security-Policy";
    return function (req, res) {
        const nonce = nonceB64url(16);
        if (!nonce) {
            res.status(500);
            res.json({ error: "csp: nonce generation failed" });
            return 1;
        }
        req.ctx = req.ctx || {};
        req.ctx.csp_nonce = nonce;
        res.header(headerName, buildHeader(profile, nonce));
        return 0;
    };
}

/**
 * Pico-compatible CSP profile. Allows inline `style="..."` attributes
 * (Pico v2 classless uses them on `<details>`, `<dialog>`, and some
 * form controls). `<script>` and `<style>` blocks still require the
 * per-request nonce.
 *
 * @param {{ reportOnly?: boolean }} [opts]
 * @returns {Function} Middleware suitable for `app.use("*", "/*", mw)`.
 */
function htmx(opts) {
    return makeMiddleware("htmx", opts);
}

/**
 * Strict CSP profile (no inline-style escape). Blocks all inline
 * styles including `style="..."` attributes. Use when you control
 * every template and ship a CSS framework without inline styles.
 *
 * @param {{ reportOnly?: boolean }} [opts]
 * @returns {Function}
 */
function strict(opts) {
    return makeMiddleware("strict", opts);
}

/**
 * Generate a fresh CSP nonce (base64url, 128-bit, unpadded). Useful
 * for callers that want to compose their own CSP middleware but reuse
 * Hull's nonce generation. Returns null on crypto.random failure.
 *
 * @returns {string|null}
 */
function nonce() {
    return nonceB64url(16);
}

export const csp = { htmx, strict, nonce };
export default csp;

File: stdlib/lua/hull/web/middleware/csp.lua
--- Content-Security-Policy middleware factory with per-request nonce.
--
-- @module hull.web.middleware.csp
-- @license AGPL-3.0-or-later
--
-- Generates a fresh 128-bit nonce on every request, exposes it via
-- `req.ctx.csp_nonce`, and sets the `Content-Security-Policy` header
-- on the outbound response. Templates that need a nonce read it from
-- the context and emit it on `<script>` and `<style>` tags:
--
--     <script nonce="{{ csp_nonce }}">...</script>
--     <style nonce="{{ csp_nonce }}">...</style>
--
-- Three named profiles are provided. All three include the same
-- non-nonce directives (default-src 'self', img-src 'self' data:,
-- form-action 'self', frame-ancestors 'none', base-uri 'self') and
-- differ only in how strictly they treat inline styles:
--
-- * `csp.htmx()`. The Pico-compatible profile. Allows inline HTML
--   `style="…"` attributes via `style-src-attr 'unsafe-inline'`.
--   `<script>` and `<style>` blocks still require the per-request
--   nonce. Use this with Pico v2 classless (which uses inline styles
--   on some components: `<details>`, `<dialog>`, form controls).
-- * `csp.strict()`. The strictest profile. No `style-src-attr`
--   escape; inline styles are blocked entirely. Use when you control
--   every template + ship a CSS framework without inline styles.
-- * `csp.custom(opts)`. Build your own. See `opts` shape below.
--
-- All three middlewares set the header on the response and continue
-- (return 0). Failure modes are limited to `crypto.random` failure,
-- which is treated as a server error (return 1, 500 response).

local crypto = require("hull.crypto")

local csp = {}

-- Base64url-encode raw bytes for use as a CSP nonce attribute value.
-- 16 bytes (128 bits of entropy) -> 22 chars unpadded. The base64url
-- alphabet is HTML-attribute-safe. Strip trailing '='.
local function nonce_b64url(n_bytes)
    n_bytes = n_bytes or 16
    local raw = crypto.random(n_bytes)
    if not raw or #raw ~= n_bytes then return nil end
    local b64 = crypto.base64url_encode(raw)
    if not b64 then return nil end
    -- crypto.base64url_encode already strips padding in Hull's
    -- implementation; this is defensive in case it changes.
    return (b64:gsub("=+$", ""))
end

-- Build the CSP header value for a given profile + nonce.
local function build_header(profile, nonce)
    local nonce_token = "'nonce-" .. nonce .. "'"
    local parts = {
        "default-src 'self'",
        "script-src 'self' " .. nonce_token,
        "style-src 'self' " .. nonce_token,
        "img-src 'self' data:",
        "form-action 'self'",
        "frame-ancestors 'none'",
        "base-uri 'self'",
    }
    if profile == "htmx" then
        -- Pico v2 classless uses inline style="…" attributes on a few
        -- components. style-src-attr controls ONLY inline style
        -- attributes; <style> blocks still require the nonce above.
        table.insert(parts, "style-src-attr 'unsafe-inline'")
    end
    return table.concat(parts, "; ")
end

-- The middleware body. Shared by all profiles; differs only on
-- `profile` arg passed in by the factory wrapper.
local function make_middleware(profile, opts)
    opts = opts or {}
    local header_name = opts.report_only and
        "Content-Security-Policy-Report-Only" or
        "Content-Security-Policy"
    return function(req, res)
        local nonce = nonce_b64url(16)
        if not nonce then
            res:status(500)
            res:json({ error = "csp: nonce generation failed" })
            return 1
        end
        req.ctx = req.ctx or {}
        req.ctx.csp_nonce = nonce
        res:header(header_name, build_header(profile, nonce))
        return 0
    end
end

--- Pico-compatible CSP profile.
--
-- Allows inline `style="…"` attributes (Pico v2 classless uses them
-- on `<details>`, `<dialog>`, and a few form controls). `<script>`
-- and `<style>` blocks still require the per-request nonce.
--
-- @tparam ?{report_only=boolean} opts
-- @treturn function  Middleware suitable for `app.use("*", "/*", mw)`.
function csp.htmx(opts)
    return make_middleware("htmx", opts)
end

--- Strict CSP profile (no inline-style escape).
--
-- Blocks all inline styles, including `style="…"` attributes. Use
-- when you control every template and ship a CSS framework without
-- inline styles.
--
-- @tparam ?{report_only=boolean} opts
-- @treturn function
function csp.strict(opts)
    return make_middleware("strict", opts)
end

--- Generate a fresh CSP nonce (base64url, 128-bit, unpadded).
--
-- Exposed for callers that want to compose their own CSP middleware
-- but reuse Hull's nonce generation. Returns nil on `crypto.random`
-- failure.
--
-- @treturn string|nil
function csp.nonce()
    return nonce_b64url(16)
end

return csp
